File: infrastructure/src/main/java/com/okaapi/kotlindemo/adapters/PostsPersistenceAdapter.java
package com.okaapi.kotlindemo.adapters;

import com.okaapi.kotlindemo.model.Post;
import com.okaapi.kotlindemo.ports.PostsPersistencePort;
import com.okaapi.kotlindemo.repositories.PostsRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class PostsPersistenceAdapter implements PostsPersistencePort {

    private final PostsRepository postsRepository;

    public PostsPersistenceAdapter(PostsRepository postsRepository) {
        this.postsRepository = postsRepository;
    }

    @Override
    public List<Post> getPosts() {
        return postsRepository.findAll()
                .stream()
                .map(postEntity -> new Post(postEntity.getTitle(), postEntity.getContent()))
                .collect(Collectors.toList());
    }
}

File: infrastructure/src/main/kotlin/com/okaapi/kotlindemo/adapters/PostsPersistenceAdapter.kt
package com.okaapi.kotlindemo.adapters

import com.okaapi.kotlindemo.model.Post
import com.okaapi.kotlindemo.model.PostEntity
import com.okaapi.kotlindemo.ports.PostsPersistencePort
import com.okaapi.kotlindemo.repositories.PostsRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.reactive.asFlow
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux

@Service
class PostsPersistenceAdapter(private val postsRepository: PostsRepository) : PostsPersistencePort {
    override suspend fun getPosts(): Flow<Post> {
        return postsRepository.findAll().asFlow().map(PostEntity::toPost)
    }
}